File: leetcode/RotateImage/rotate.java
public class Solution {
    public void rotate(int[][] matrix) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int n = matrix.length;
        //if (n == 0) return;
        int layer = n / 2;
        for (int i = 0; i < layer; i++) {
            for (int j = i; j < n - 1 - i; j++) {
                int x = matrix[i][j];
                matrix[i][j] = matrix[n - 1 - j][i];
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];
                matrix[j][n - 1 - i] = x;
            }
        }
    }
}

File: leetcode/RotateArray/rotate.cpp
class Solution {
public:
    void reverse(vector<int>& nums, int i, int j) {
        if (i >= j) return;
        for (int k = 0; k < (j - i) / 2; k++) {
            int tmp = nums[k + i];
            nums[k + i] = nums[j - 1 - k];
            nums[j - 1 - k] = tmp;
        }
    }
    void rotate(vector<int>& nums, int k) {
        k %= nums.size();
        reverse(nums, 0, nums.size() - k);
        reverse(nums, nums.size() - k, nums.size());
        reverse(nums, 0, nums.size());
    }
};
